File: server_python/routes/pnr.py
# ==========================================
# AirOps AI — PNR API Route
# ==========================================

from fastapi import APIRouter
from services.auth_engine import secure_pnr_lookup
from services.decision_engine import evaluate_flight_disruption, evaluate_refund_eligibility, evaluate_baggage_rights

router = APIRouter(prefix="/api/pnr", tags=["PNR"])


@router.get("/{locator}")
def get_pnr(locator: str):
    result = secure_pnr_lookup(locator.upper())
    if not result["found"]:
        return {"error": result["error"]}, 404
    return result["pnr"]


@router.get("/{locator}/rights")
def get_pnr_rights(locator: str):
    result = secure_pnr_lookup(locator.upper())
    if not result["found"]:
        return {"error": result["error"]}, 404

    pnr = result["pnr"]
    flight_rights = []

    for seg in pnr.get("segments", []):
        is_cancelled = seg.get("segment_status") == "cancelled"
        delay_min = seg.get("delay_minutes", 0) or 0
        disruption = evaluate_flight_disruption(delay_min, is_cancelled, False)
        flight_rights.append({"flightNumber": seg["flight_number"], "origin": seg["origin"], "destination": seg["destination"], **disruption})

    baggage_rights = []
    for b in pnr.get("baggage", []):
        if b.get("baggage_status") in ("missing", "delayed"):
            baggage_rights.append(evaluate_baggage_rights(True, b.get("days_missing", 0)))

    refund_eligibility = None
    tickets = pnr.get("tickets", [])
    if tickets:
        t = tickets[0]
        has_irop = any(s.get("segment_status") == "cancelled" or (s.get("delay_minutes") or 0) >= 240 for s in pnr.get("segments", []))
        refund_eligibility = evaluate_refund_eligibility(
            {"refundable": bool(t.get("refundable")), "change_fee": t.get("change_fee", 0), "cancel_fee": t.get("cancel_fee", 0), "voucher_allowed": bool(t.get("voucher_allowed")), "fare_family": t.get("fare_family", "basic")},
            "involuntary" if has_irop else "voluntary",
            t.get("total_amount", 0),
        )

    return {"locator": locator.upper(), "flightRights": flight_rights, "baggageRights": baggage_rights, "refundEligibility": refund_eligibility}

File: server/src/routes/pnr.ts
// ==========================================
// AirOps AI — PNR API Route
// ==========================================
import { Router } from 'express';
import { securePnrLookup } from '../services/authEngine.js';
import { evaluateFlightDisruption, evaluateRefundEligibility, evaluateBaggageRights } from '../services/decisionEngine.js';
import { db } from '../db/sqlite.js';

const router = Router();

// GET /api/pnr/:locator
router.get('/:locator', (req, res) => {
  try {
    const { locator } = req.params;
    const result = securePnrLookup(locator.toUpperCase());
    if (!result.found) {
      res.status(404).json({ error: result.error });
      return;
    }
    res.json(result.pnr);
  } catch (error) {
    console.error('PNR lookup error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// GET /api/pnr/:locator/rights — ANAC rights for current flight status
router.get('/:locator/rights', (req, res) => {
  try {
    const { locator } = req.params;
    const result = securePnrLookup(locator.toUpperCase());
    if (!result.found) {
      res.status(404).json({ error: result.error });
      return;
    }

    const pnr = result.pnr;
    const rights = [];

    for (const seg of pnr.segments || []) {
      const isCancelled = seg.segment_status === 'cancelled';
      const delayMin = seg.delay_minutes || 0;
      const disruption = evaluateFlightDisruption(delayMin, isCancelled, false);
      rights.push({ flightNumber: seg.flight_number, origin: seg.origin, destination: seg.destination, ...disruption });
    }

    // Baggage rights
    const baggageRights = (pnr.baggage || []).filter((b: any) => b.baggage_status === 'missing' || b.baggage_status === 'delayed').map((b: any) => {
      return evaluateBaggageRights(true, b.days_missing || 0);
    });

    // Refund eligibility
    let refundEligibility = null;
    if (pnr.tickets?.length > 0) {
      const ticket = pnr.tickets[0];
      const hasIROP = pnr.segments?.some((s: any) => s.segment_status === 'cancelled' || s.delay_minutes >= 240);
      refundEligibility = evaluateRefundEligibility(
        { refundable: !!ticket.refundable, change_fee: ticket.change_fee || 0, cancel_fee: ticket.cancel_fee || 0, voucher_allowed: !!ticket.voucher_allowed, fare_family: ticket.fare_family || 'basic' },
        hasIROP ? 'involuntary' : 'voluntary',
        ticket.total_amount || 0
      );
    }

    res.json({ locator, flightRights: rights, baggageRights, refundEligibility });
  } catch (error) {
    console.error('Rights evaluation error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;
